packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -6,372 +6,188 @@
+        ) revert L1_INVALID_EVIDENCE();
+
+        TaikoData.ProposedBlock storage proposal = state.blocks[
+            meta.id % config.maxNumBlocks
+        ];
 
-        if (FLAG_VALIDATE_HEADER_FOR_METADATA) {
-            if (
-                evidence.header.parentHash == 0 ||
-                evidence.header.beneficiary != evidence.meta.beneficiary ||
-                evidence.header.difficulty != 0 ||
-                evidence.header.gasLimit !=
-                evidence.meta.gasLimit + config.anchorTxGasLimit ||
-                evidence.header.gasUsed == 0 ||
-                evidence.header.timestamp != evidence.meta.timestamp ||
-                evidence.header.extraData.length !=
-                evidence.meta.extraData.length ||
-                keccak256(evidence.header.extraData) !=
-                keccak256(evidence.meta.extraData) ||
-                evidence.header.mixHash != evidence.meta.mixHash
-            ) revert L1_META_MISMATCH();
-        }
+        if (proposal.metaHash != LibUtils.hashMetadata(meta))
+            revert L1_EVIDENCE_MISMATCH();
 
-        // For alpha-2 testnet, the network allows any address to submit ZKP,
-        // but a special prover can skip ZKP verification if the ZKP is empty.
+        TaikoData.ForkChoice storage fc = state.forkChoices[
+            blockId % config.maxNumBlocks
+        ][proposal.nextForkChoiceId];
 
         bool oracleProving;
+        if (uint256(fc.chainData.blockHash) <= 1) {
+            // 0 or 1 (placeholder) indicate this block has not been proven
+            if (config.enableOracleProver) {
+                if (msg.sender != resolver.resolve("oracle_prover", false))
+                    revert L1_NOT_ORACLE_PROVER();
 
-        TaikoData.ForkChoice storage fc = state.forkChoices[target.id][
-            evidence.header.parentHash
-        ];
+                oracleProving = true;
+            }
 
-        bytes32 blockHash = evidence.header.hashBlockHeader();
-        bytes32 _blockHash = blockHashOverride == 0
-            ? blockHash
-            : blockHashOverride;
+            fc.chainData = ChainData(evidence.blockHash, evidence.signalRoot);
 
-        if (fc.blockHash == 0) {
-            address oracleProver = resolver.resolve("oracle_prover", true);
-            if (msg.sender == oracleProver) {
-                oracleProving = true;
+            if (oracleProving) {
+                // make sure we reset the prover address to indicate it is
+                // proven by the oracle prover
+                fc.prover = address(0);
             } else {
-                if (oracleProver != address(0)) revert L1_NOT_ORACLE_PROVER();
                 fc.prover = evidence.prover;
                 fc.provenAt = uint64(block.timestamp);
             }
-            fc.blockHash = _blockHash;
         } else {
-            if (fc.blockHash != _blockHash) revert L1_CONFLICT_PROOF();
             if (fc.prover != address(0)) revert L1_ALREADY_PROVEN();
```

Comment:
How the `if (fc.prover != address(0)) revert L1_ALREADY_PROVEN()` can be true, if the `fc` is every time a new instance ?

Reply:
It is important to realize that `fc` is not a new instance, it is a reusable in-storage struct. Its values are reset (to place-holding values) after the corresponding block is verified.

<img width="539" alt="Screenshot 2023-03-22 at 09 11 53" src="https://user-images.githubusercontent.com/99078276/226776307-24d849b6-c07e-42ae-b80f-559116042bb3.png">

But here the reason `fc.prover` can be non-empty is still because more than one regular provers can attempt to prove the very same fork choice, after the first proving is success, this value is non-zero address, to prevent other provers to prove the same fork choice again.